File: lua/05-functions/recursion.lua
-- Factorial: n! = n * (n-1) * ... * 1
local function factorial(n)
    if n <= 1 then
        return 1
    end
    return n * factorial(n - 1)
end

print("5! =", factorial(5))
print("10! =", factorial(10))

-- Fibonacci sequence via recursion
local function fib(n)
    if n < 2 then
        return n
    end
    return fib(n - 1) + fib(n - 2)
end

for i = 0, 9 do
    io.write(fib(i), " ")
end
print()

File: pascal/05-functions/recursion.pas
program Recursion;

{ n! = n * (n-1)! with a base case of 1 }
function Factorial(n: Integer): Int64;
begin
    if n <= 1 then
        Factorial := 1
    else
        Factorial := n * Factorial(n - 1);
end;

{ Each Fibonacci number is the sum of the previous two }
function Fibonacci(n: Integer): Integer;
begin
    if n < 2 then
        Fibonacci := n
    else
        Fibonacci := Fibonacci(n - 1) + Fibonacci(n - 2);
end;

var
    i: Integer;

begin
    WriteLn('Factorials:');
    for i := 1 to 10 do
        WriteLn(i, '! = ', Factorial(i));

    WriteLn;
    Write('Fibonacci sequence: ');
    for i := 0 to 10 do
        Write(Fibonacci(i), ' ');
    WriteLn;
end.
